[MSHADE-307] adding useDefaultConfiguration flag in ShadeMojo

rmannibucau opened the pull request:
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSHADE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MSHADE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHADE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

- [X] I'm already an ASF committer

rmannibucau commented:
Common sounds a good alternative but too generic, I thought about "auto" but was not that happy too. Maybe "autoConfigure" + adding a spi on transformers and filters would make more sense?

eolivelli commented:
autoConfigure + SPI will be great, but it a more complex change.

The other way would be to have such "useDefaultConfiguration", turned *on* by default, but with a bump in the major version for the plugin.

@rfscholte  what do you think ?

rfscholte commented:
In general I'd like prefer to keep the number of parameters to a minimum. So if we want to do this, why not set these as defaults. When there are explicit filters/transformers, these will be used. Any empty list would mean: no filters or transformers. And this implies there should be an explicit filter for signatures.

rfscholte commented:
The new version will be 3.n+1.0. The first 3 means that is requires at least Maven 3. We discovered that this pattern makes is very clear when certain plugins can be used.
We should add this change to the index.html file too, just to be clear. (in the end the result should be the same, because if filters and transformers are specified, nothing will change)

rmannibucau commented:
Hmm, i thought of that - active by default and ignored if set by the user - but it is rare it is sufficient and does need a log4j, openwebbeans or other transformer so i preferred the composition. Should we have it with 3 values: default = use, skip and merge?

eolivelli commented:
Okay for versioning and index.html
I can take of take (with a little guidance).

For use/skip/merge....things are getting complicated and complicated.

We should default to 'merge', but will 'use' be useful? As 'merge' is a super set of 'use'

rmannibucau commented:
Hmm it must be combinable or it is useless and config stays complex.

Merge (default), override, skip?

rfscholte commented:
Sorry, but it will be default or override: easy to understand and maximum control. If you want full merge/override-control, you need to do that with the magic attributes as described at https://blog.sonatype.com/2011/01/maven-how-to-merging-plugin-configuration-in-complex-projects/

rmannibucau commented:
@rfscholte this is my intention but here it does not work cause the config is implicit, this is why i proposed 3 states. Now if you say the impl is ok to turn on by default it can be enough for the next release (and we enhance later if needed). If so, do you have a name preference?

rfscholte commented:
> If so, do you have a name preference?

for what?

rmannibucau commented:
Guys, I added the SPI and removed the toggle, can you have a look and let me know if it looks better this way? I like this solution cause we can plug the SPI in log4j, openwebbeans, and other extensions

rfscholte commented on src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:
```diff
@@ -762,55 +766,72 @@ private File resolveArtifactSources( Artifact artifact )
         if ( transformers == null )
         {
-            return Collections.emptyList();
+            return resourceTransformers;
```
You're still merging. Here you should replace `Collections.emptyList()` with `getDefaultResourceTransformers()`

rmannibucau replied:
this is the case, resourceTransformers is initialized with getDefaultResourceTransformers. If the else skips the default transformers the feature is pointless since in 80% of the case you need to add a custom transformer or customize a config (like mainClass for the manifest)

rfscholte replied:
Let's agree to disagree

rmannibucau replied:
I'm fine to disagree but I'd like to solve the issue which aims to reduce the entry cost of the plugin and save hours of debug time due to a bad packaging which is hard to identify - and even more with jpms since module-info.java drops SPI registration :s.

any alternative proposal to have a trivial to enable good auto configuration compatible with user customizations?

rfscholte replied:
Can't think of a clean, clear and simple solution yet. So let's split it into 2 issues: 1. have good default transformers; 2. combine default+custom transfomers

rmannibucau replied:
Ok, let's split. Just to ensure I get it and not do work for nothing, it means 3 PR - no order:

1. defaults (transformers and filters)
2. SPI
3. combine
?

rmannibucau replied:
1 done at https://github.com/apache/maven-shade-plugin/pull/12

rfscholte commented on src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:
```diff
@@ -762,55 +766,72 @@ private File resolveArtifactSources( Artifact artifact )
 
-        return Arrays.asList( transformers );
+    private List<ResourceTransformer> getDefaultResourceTransformers()
+    {
```
Since we're not merging, so you can return a list with these 2 transformers.
Also document this on the `transfomers`-parameter.

rmannibucau replied:
guess it is addressed by previous comment, we want to merge ;)

rfscholte commented on src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:
```diff
@@ -823,6 +844,50 @@ private File resolveArtifactSources( Artifact artifact )
         }
 
+        // first check for backward compatibility this is not already configured explicitly
+        boolean addExclusion = true;
```
We need a similar trick here:

    public Collection<Filter> getFilters()
    {
      if ( this.filters == null )
      {
        return defaultFilters();
      }
      else
      {
        return this.filters;
      }
    }

rmannibucau replied:
likely the same

rfscholte commented on src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java:
```diff
@@ -823,6 +844,50 @@ private File resolveArtifactSources( Artifact artifact )
+                    new HashSet<String>( Arrays.asList( "META-INF/*.SF", "META-INF/*.DSA", "META-INF/*.RSA" ) ) ) );
+        }
+
+        for ( final Filter filter : ServiceLoader.load( Filter.class ) )
```
Using ServiceLoader requires a separate JIRA issue, and I wonder if this is going to work. See https://issues.apache.org/jira/browse/MNG-6275

jira-importer commented:
Resolve #308

jira-importer commented:
Resolve #308
